src/libgit2/repository.c

```diff
@@ -3246,14 +3246,18 @@ int git_repository_set_workdir(
 	if (git_fs_path_prettify_dir(&path, workdir, NULL) < 0)
 		return -1;
 
-	if (repo->workdir && strcmp(repo->workdir, path.ptr) == 0)
```

Comment:
Just wanted to propose ad additional optimisation: I don't see that much sense in allocating a string and potentially to deallocate it. I think we should do it  if and only if `workdir != repo->workdir` (namely we should call `git_fs_path_prettify_dir` after checking that we really have a new different working dir).

```
if (!repo->workdir || strcmp(repo->workdir, wordir) == 0)
  return 0;
if (git_fs_path_prettify_dir(&path, workdir, NULL) < 0)
  return -1;
```

Reply:
This is not logically equivalent, though. Currently, we're comparing the canonicalized input against the existing workdir (which is also already canonicalized). If we do the comparison before the canonicalization (`prettify`), then if `repo->workdir = "/path/to/current/workdir/"` and `set_workdir` was called with `/path/to/current/workdir`, then the `strcmp` would not be equivalent, and we'd go through the whole "change the current workdir" logic incorrectly.

Reply:
Oh, I see! Sorry again (I didn't check too much in detail what `git_fs_path_prettify_dir` was really doing).

Reply:
No worries at all - I appreciate the extra set of eyes!